src/tests/script_admin_json.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file script_admin_json.cpp Tests for the Squirrel -> JSON conversion in ScriptAdmin. */
+
+#include "../stdafx.h"
+
+#include "../3rdparty/catch2/catch.hpp"
+
+#include "../network/network_admin.h"
+#include "../game/game_instance.hpp"
+#include "../script/api/script_object.hpp"
+#include "../script/squirrel.hpp"
+#include "../script/squirrel_std.hpp"
+#include "../script/squirrel_class.hpp"
+#include "../3rdparty/fmt/format.h"
+
+#include <squirrel.h>
+
+class TestServerNetworkAdminSocketHandler : public ServerNetworkAdminSocketHandler {
+public:
+	std::string expected_json{};
+	int expected_call{0};
+
+	TestServerNetworkAdminSocketHandler() : ServerNetworkAdminSocketHandler(0)
+	{
+		status = ADMIN_STATUS_ACTIVE;
+		update_frequency[ADMIN_UPDATE_GAMESCRIPT] = ADMIN_FREQUENCY_AUTOMATIC;
+	}
+
+	NetworkRecvStatus SendGameScript(const std::string_view json) override
+	{
+		CHECK(json == expected_json);
+		expected_call--;
+
+		return NETWORK_RECV_STATUS_OKAY;
+	}
+};
+
+class TestGameInstance: public GameInstance {
+public:
+	ScriptObject::ActiveInstance active;
+
+	TestGameInstance() : GameInstance(), active(this)
+	{
+	}
+};
+
+static void TestJSON(TestServerNetworkAdminSocketHandler *admin, Squirrel *engine, const std::string &test_script, const std::string &expected_json)
+{
+	admin->expected_json = expected_json;
+	admin->expected_call = expected_json.empty() ? 0 : 1;
+
+	/* Load and run the script */
+	sq_pushroottable(engine->GetVM());
+	sq_compilebuffer(engine->GetVM(), test_script.c_str(), test_script.size(), "test", SQTrue);
+	sq_push(engine->GetVM(), -2);
+	sq_call(engine->GetVM(), 1, SQFalse, SQTrue);
+	sq_pop(engine->GetVM(), 1);
+
+	CHECK(admin->expected_call == 0);
+}
+
+TEST_CASE("Squirrel -> JSON conversion")
+{
+	/* Create a fake admin socket that can validate our JSON. */
+	ServerNetworkAdminSocketHandler::CanAllocateItem();
+	TestServerNetworkAdminSocketHandler *a = new TestServerNetworkAdminSocketHandler();
```

Comment:
I would use `std::unique_ptr`. Then you don't need to `delete` is at the end of the test.